source/agora/consensus/EnrollmentManager.d

```diff
@@ -168,6 +183,90 @@ public class EnrollmentManager
         Utils.getCollectorRegistry().addCollector(&this.collectValidatorStats);
     }
 
+    /***************************************************************************
+
+        Update the validator key index maps for the provided height
+
+    ***************************************************************************/
+
+    public void updateValidatorIndexMaps (Height height) @safe
+    {
+        PublicKey[] keys;
+        if (!this.getEnrolledPublicKeys(keys))
+            assert(0);  // should not happen
+
+        assert(keys.length > 0);
+        import std.algorithm : each, sort;
+
+        static PublicKey[] keys_arr;
+        keys_arr.length = 0;
+        () @trusted { assumeSafeAppend(keys_arr); }();
+        foreach (ref PublicKey k; keys)
+            keys_arr ~= k;
+        keys_arr.sort();
+
+        foreach (idx, key; keys_arr)
+        {
+            log.trace("Set index {} for key_to_index[{}] for {}", idx, height, PublicKey(key[]));
+            const K = Point(key[]);
+            this.key_to_index[height][K] = idx;
+            log.trace("Set key {} for index_to_key[{}] for {}", PublicKey(key[]), height, idx);
+            this.index_to_key[height][idx] = K;
+        }
+    }
+
+    /***************************************************************************
+
+        Params:
+            height = the height at which to look up the mapping for
+            K = the Key for which to find the bitfield index to use
+
+        Returns:
+            the count of validators at this enrollment height
+
+    ***************************************************************************/
+
+    public size_t getCountOfValidators (Height height) nothrow @safe
+    {
+        if (height >= this.key_to_index.length)
+            return 0;
+        return this.key_to_index[Height(height - 1)].length;
+    }
+
+    /***************************************************************************
+
+        Params:
+            height = the height at which to look up the mapping for
+            K = the Key for which to find the bitfield index to use
+
+        Returns:
+            the index of this key, or ulong.max if none was found
+
+    ***************************************************************************/
+
+    public ulong getIndexOfValidator (Height height, Point K) nothrow @safe
+    {
+        scope (failure) return ulong.max;
```

Comment:
This is very curious that it works. Usually `scope (failure)` catches and re-throws any exception that is in-flight. But here it seems to circumvent it with a `return`.

I'm not sure if this is part of the design, or just accidentally works correctly..

@Geod24 did you know about this feature?

Reply:
Actually I am pretty sure this is just buggy behavior. I'm filing a bug for it now, because it breaks safety.

Reply:
https://issues.dlang.org/show_bug.cgi?id=21443

Reply:
I'll replace with a try / catch.

Reply:
Updated